cli/connector/eval_plain_text.go

```diff
@@ -0,0 +1,389 @@
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/vmihailenco/msgpack/v5"
+
+	"github.com/tarantool/cartridge-cli/cli/templates"
+	"gopkg.in/yaml.v2"
+)
+
+const (
+	endOfYAMLOutput = "\n...\n"
+	endOfLuaOutput  = ";"
+
+	tagPushPrefixYAML = `%TAG !push!`
+	tagPushPrefixLua  = `-- Push`
+)
+
+type EvalPlainTextOpts struct {
+	ReadTimeout  time.Duration
+	PushCallback func(interface{})
+	ResData      interface{}
+}
+
+type PlainTextEvalRes struct {
+	Success bool   `yaml:"success"`
+	DataEnc string `yaml:"data_enc"`
+	ErrStr  string `yaml:"err"`
+}
+
+// callPlainTextConnYAML calls function on Tarantool instance
+// Function should return `interface{}`, `string` (res, err)
+// to be correctly processed.
+// Processes only YAML output.
+func callPlainTextConnYAML(conn net.Conn, funcName string, args []interface{}, opts EvalPlainTextOpts) ([]interface{}, error) {
+	evalFunc, err := templates.GetTemplatedStr(&callFuncTmpl, map[string]string{
+		"FunctionName": funcName,
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("Failed to instantiate call function template: %s", err)
+	}
+
+	return evalPlainTextConnYAML(conn, evalFunc, args, opts)
+}
+
+// evalPlainTextConnYAML calls function on Tarantool instance
+// Function should return `interface{}`, `string` (res, err)
+// to be correctly processed.
+// Processes only YAML output.
+func evalPlainTextConnYAML(conn net.Conn, funcBody string, args []interface{}, opts EvalPlainTextOpts) ([]interface{}, error) {
+	if err := formatAndSendEvalFunc(conn, funcBody, args, evalFuncYAMLTmpl); err != nil {
+		return nil, err
+	}
+
+	// recv from socket
+	resBytes, err := readFromPlainTextConnYAML(conn, opts)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to check returned data: %s", err)
+	}
+
+	data, err := processEvalTarantoolResYAML(resBytes, opts.ResData)
+	if err != nil {
+		return nil, err
+	}
+
+	return data, nil
+}
+
+// func evalPlainTextConnLua(conn net.Conn, funcBody string, args []interface{}, opts EvalPlainTextOpts) (interface{}, error) {
+// 	if err := formatAndSendEvalFunc(conn, funcBody, args, evalFuncYAMLTmpl); err != nil {
+// 		return nil, err
+// 	}
+
+// 	// recv from socket
+// 	resBytes, err := readFromPlainTextConnLua(conn, opts)
+// 	if err != nil {
+// 		return nil, fmt.Errorf("Failed to check returned data: %s", err)
+// 	}
+
+// 	res, err := processEvalTarantoolResLua(resBytes)
+// 	if err != nil {
+// 		return nil, fmt.Errorf("Failed to parse data returned from instance: %s", err)
+// 	}
+
+// 	if !res.Success {
+// 		return nil, fmt.Errorf(res.ErrStr)
+// 	}
+
+// 	return res.Data, nil
+// }
+
+func formatAndSendEvalFunc(conn net.Conn, funcBody string, args []interface{}, evalFuncTmpl string) error {
+	if args == nil {
+		args = []interface{}{}
+	}
+
+	argsEncoded, err := msgpack.Marshal(args)
+	if err != nil {
+		return fmt.Errorf("Failed to encode args: %s", err)
+	}
+
+	evalFunc, err := templates.GetTemplatedStr(&evalFuncTmpl, map[string]string{
```

Comment:
any "eval" could be considered as call('package.loaded.console.eval', 'string'). Do we need evals?

Reply:
We use plain text protocol, there wouldn't be any beautiful `call`.